packages/flutter/lib/src/rendering/paragraph.dart

```diff
@@ -1616,15 +1620,32 @@ class _SelectableFragment with Selectable, ChangeNotifier implements TextLayoutM
     return result;
   }
 
-  TextPosition _getNextPosition(TextBoundary boundary, TextPosition position, bool forward) {
-    if (forward) {
-      return _clampTextPosition(
-        (PushTextPosition.forward + boundary).getTrailingTextBoundaryAt(position)
-      );
+  TextPosition _beyondTextBoundary(TextPosition extent, bool forward, TextBoundary textBoundary) {
+    final int newOffset = forward
+      ? textBoundary.getTrailingTextBoundaryAt(extent.offset) ?? range.end
+      : textBoundary.getLeadingTextBoundaryAt(extent.offset - 1) ?? range.start;
+    return TextPosition(offset: newOffset);
+  }
+
+  TextPosition _toTextBoundary(TextPosition extent, bool forward, TextBoundary textBoundary) {
+    assert(extent.offset >= 0);
+    final int caretOffset;
+    switch (extent.affinity) {
+      case TextAffinity.upstream:
+        if (extent.offset < 1 && !forward) {
+          assert (extent.offset == 0);
+          return const TextPosition(offset: 0);
+        }
+        caretOffset = math.max(0, extent.offset - 1);
+        break;
+      case TextAffinity.downstream:
+        caretOffset = extent.offset;
+        break;
     }
-    return _clampTextPosition(
-      (PushTextPosition.backward + boundary).getLeadingTextBoundaryAt(position),
-    );
+    final int offset = forward
```

Comment:
can we reuse `_beyondTextBoundary` here?

Reply:
It's different, the beyond one has a `-1` somewhere.

Reply:
sounds good